### PR TITLE
Add sampling observer auxiliary output hooks

### PR DESCRIPTION
**Pull request by alecsolder**
## Motivation

Some sampling-time integrations need to observe logits before grammar masking and attach structured per-token metadata to generated responses. Doing this currently requires coupling the integration directly to model-runner and scheduler internals.

## Modifications

- Add an optional sampling observer interface around grammar masking and token sampling.
- Carry observer-owned auxiliary tensors through asynchronous result copying and pipeline-parallel transport.
- Consume committed-token metadata after scheduler output processing and expose accumulated values through the existing customized response metadata path.
- Preserve the existing sampling path when no observer is configured.

## Accuracy Tests

- `pytest -q test/registered/unit/managers/test_batch_result_processor_hidden_states.py test/registered/unit/managers/test_generation_auxiliary_output.py test/registered/unit/managers/test_output_streamer_customized_info.py test/registered/unit/sampling/test_sampling_batch_info.py`
  - 75 passed, 2 subtests passed
- The tests include exact numerical-equivalence coverage for the observer-disabled logits path.

## Speed Tests and Profiling

No standalone benchmark was run. When no observer is configured, the added path performs no auxiliary tensor computation or transfer.

## Checklist

- [x] Format the code with the repository pre-commit hooks.
- [x] Add registered unit tests for the new behavior.
- [x] Document the extension contracts in the protocol docstrings.
- [x] Verify numerical identity when the observer is disabled.
- [x] Keep the change independent of model-specific implementations.

## Original commits

- `60d0a82433`
- `df3183df41`
- `46e1b8eca2`
- `2f598043c7`

---
### CI States

Latest PR Test (Base): :x: [Run #32492449054](https://github.com/sgl-project/sglang/actions/runs/32492449054)
Latest PR Test (Extra): :x: [Run #32492448997](https://github.com/sgl-project/sglang/actions/runs/32492448997)
Latest PR Test (AMD ROCm 7.2): :hourglass_flowing_sand: [Run #32492449051](https://github.com/sgl-project/sglang/actions/runs/32492449051)

### CODE REVIEW

File: python/sglang/srt/sampling/sampling_observer.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        commits: Sequence[Optional[CommittedTokens]],
+    ) -> None: ...
+
+
+class SamplingObserver(Protocol):
+    """Invocation-scoped hooks around the production grammar mask and sampler.
+
+    Returning ``None`` from ``before_grammar`` skips ``after_sample``.
+    These hooks are driven by ``ModelRunner.sample``. Specialized speculative
+    workers that sample elsewhere must produce their own auxiliary output.
```
Comment (Ubospica) [inline python/sglang/srt/sampling/sampling_observer.py]: When a speculative algorithm is enabled, sampling happens in the spec worker and `ModelRunner.sample` is bypassed, so an installed observer silently produces no auxiliary output (same for dLLM batches, whose result processor never calls `consume_auxiliary_output`). Relying on this docstring makes the failure mode silent data loss on the response side.

Since `SchedulerOutputStreamer.__post_init__` already fails fast for the unsupported Rust-egress combination, consider an analogous startup check: raise when an observer is installed together with a speculative algorithm (or any configuration whose sampling path does not drive these hooks).

File: python/sglang/srt/managers/utils.py [hunk 2]
```diff
@@ -170,8 +173,17 @@ def copy_to_cpu(self, return_logprob: bool, return_hidden_states: bool = True):
...
+        self.copy_auxiliary_output_to_cpu()
+
         self.copy_done.record()
 
+    def copy_auxiliary_output_to_cpu(self) -> None:
+        if self.logits_output is None or self.auxiliary_host_output is not None:
+            return
+        device_output = self.logits_output.auxiliary_device_output
+        if device_output is not None:
+            self.auxiliary_host_output = device_output.copy_to_host(_async_d2h)
```
Comment (Ubospica) [inline python/sglang/srt/managers/utils.py]: After the host copy is enqueued, the device output stays referenced via `logits_output.auxiliary_device_output`. In overlap mode the batch result (and its `logits_output`) is kept alive by `batch_record_buf` for ~2 iterations, so the observer's device tensors are retained on GPU longer than needed — the same retention pattern that motivated clearing `sampling_info.grammar_mask` in `_preprocess_logits` and `next_token_logits` in `launch_batch_sample_if_needed`. If the observer captures anything sized like pre-grammar logits, that is a non-trivial amount of VRAM.

`_async_d2h` already `record_stream`s the source tensor, so clearing the reference here (`self.logits_output.auxiliary_device_output = None` after `copy_to_host`) looks safe on every call path: non-overlap, overlap, delayed sample, and PP first rank all enqueue the copy before the next `sample()` reassigns the field on the same thread.

`test_auxiliary_output_uses_result_copy_path_without_mutating_device_holder` asserts the field is deliberately left untouched — if there is a reason for that, a comment explaining it would help.

File: python/sglang/srt/managers/scheduler_components/output_streamer.py [hunk 3]
```diff
@@ -170,6 +180,25 @@ def _stream_output_generation(
             acc.accept(req=req)
             self._maybe_log_time_stats(req=req)
 
+        if (
+            self.has_additional_customized_info
+            and self.should_build_additional_customized_info()
+        ):
+            req_by_rid = {req.rid: req for req in reqs}
+            output_reqs = [req_by_rid[rid] for rid in acc.rids]
```
Comment (Ubospica) [inline python/sglang/srt/managers/scheduler_components/output_streamer.py]: This rebuilds a rid->req dict on every stream call and assumes rids are unique within `reqs`. Since `accept()` is already the place that decides which requests get emitted (it appends to `acc.rids`), recording the accepted `Req` objects in a parallel list on the accumulator would avoid both the per-call dict build and the rid-uniqueness assumption.

**Review by Ubospica — COMMENTED** [review]
Reviewed the full diff and traced the copy/synchronization points on each path (overlap inline `copy_to_cpu`, non-overlap forced copy with a fresh `copy_done`, PP copy-stream + `d2h_event`, pdmux split prefill). Those all look correct to me, and the `apply_logits_bias` split preserves the exact penalties -> grammar mask -> logit_bias order for the default path.

A few higher-level questions:

1. **There is no in-tree way to install any of these hooks.** `ModelRunner.sampling_observer` is initialized to `None` and never assigned anywhere; `get_output_streamer_class()`, `has_additional_customized_info`, and `update_request_meta_info` (whose new `finish_reason` parameter is unused in the base implementation) have no overriding subclass either. As written, the contracts are only exercised by unit tests that implement the protocols themselves. Could you document how an integration is expected to wire these up (subclass `ModelRunner`/`Scheduler`/`TokenizerManager` and patch construction?), or add a minimal in-tree consumer/registration point so CI actually exercises the seams?

2. **The wiring between `HostAuxiliaryOutput.consume()` and `build_additional_customized_info()` is implicit.** `consume(batch, commits)` runs right before `stream_output`, so an implementation is apparently expected to buffer per-request values keyed by `CommittedTokens.output_index` and then emit them from a subclassed streamer / tokenizer manager. Nothing in-tree demonstrates this end-to-end flow; a short example in the protocol docstrings or the PR description would make the contract much easier to evaluate.

3. The **PR Test (Extra)** run is red — worth checking whether it is related.

File: python/sglang/srt/disaggregation/prefill.py [hunk 4]
```diff
@@ -680,6 +680,12 @@ def process_batch_result_disagg_prefill(
 
         if copy_done is not None:
             copy_done.synchronize()
+        auxiliary_output = result.auxiliary_host_output
+        auxiliary_output_starts = (
+            [self.batch_result_processor._visible_output_len(req) for req in batch.reqs]
```
Comment (Ubospica) [inline python/sglang/srt/disaggregation/prefill.py]: This reaches into a private method of another component (`batch_result_processor._visible_output_len`), and the same snapshot pattern (`auxiliary_output` + `auxiliary_output_starts` guarded by `is not None`) is now duplicated in three call sites (here, `process_batch_result_prefill`, `process_batch_result_decode`). A small public helper on `SchedulerBatchResultProcessor`, e.g. `snapshot_auxiliary_output_starts(batch, result) -> Optional[List[int]]`, would remove the duplication and the private access.

File: python/sglang/srt/model_executor/model_runner.py [hunk 5]
```diff
@@ -1785,6 +1813,11 @@ def sample(
                 else forward_batch.seq_lens - 1
             ),
         )
+        if observer_state is not None:
+            logits_output.auxiliary_device_output = self.sampling_observer.after_sample(
```
Comment (Ubospica) [inline python/sglang/srt/model_executor/model_runner.py]: Nit: this uses `self.sampling_observer` while `is_active` / `before_grammar` above went through the local `observer` variable. Using the local variable is more consistent and guarantees the same object is used for the whole invocation.

File: python/sglang/srt/managers/utils.py (hunk 2 above)
Comment (Ubospica) [inline python/sglang/srt/managers/utils.py:185]: After moving to host we can release the tensor, e.g.

```suggestion
            self.auxiliary_host_output = device_output.copy_to_host(_async_d2h)
            self.logits_output.auxiliary_device_output = None
```

unless we have specific need to keep the tensor.

File: test/registered/unit/managers/test_generation_auxiliary_output.py [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+        logits_output=logits_output,
+        next_token_ids=torch.tensor([7]),
+        copy_done=CopyDone(),
+    )
+
+    result.copy_to_cpu(return_logprob=False)
+
+    assert logits_output.auxiliary_device_output is device_output
+    assert result.auxiliary_host_output is not device_output
+    assert torch.equal(result.auxiliary_host_output.values, device_output.values)
```
Comment (Ubospica) [inline test/registered/unit/managers/test_generation_auxiliary_output.py]: Following up on the change in copy_auxiliary_output_to_cpu, we may not need assert the equality here after copying.

File: python/sglang/srt/sampling/sampling_observer.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    """Scheduler-side result produced by ``DeviceAuxiliaryOutput``."""
+
+    def consume(
+        self,
+        batch: ScheduleBatch,
+        commits: Sequence[Optional[CommittedTokens]],
+    ) -> None: ...
+
+
+class SamplingObserver(Protocol):
```
Comment (Ubospica) [inline python/sglang/srt/sampling/sampling_observer.py:57]: Looks the observer is not used by the main engine yet. Maybe we can briefly introduce how to use it in the code

**Review by Ubospica — COMMENTED** [review]
The logic LGTM! Some minor comments

File: python/sglang/srt/managers/utils.py (hunk 2 above)
Comment (alecsolder) [reply python/sglang/srt/managers/utils.py:185]: Agreed, cleared right after copy

File: test/registered/unit/managers/test_generation_auxiliary_output.py (hunk 6 above)
Comment (alecsolder) [reply test/registered/unit/managers/test_generation_auxiliary_output.py]: Agreed

File: python/sglang/srt/sampling/sampling_observer.py (hunk 7 above)
Comment (alecsolder) [reply python/sglang/srt/sampling/sampling_observer.py:57]: Added documentation to explain how to register from a plugin, not going to include an example in code yet to prevent scope creep of this PR
